File: internal/cli/benchmark_test.go
package cli

import (
	"bytes"
	"encoding/json"
	"reflect"
	"strings"
	"testing"
)

func TestBenchmarkUpdateSnapshotsRejects(t *testing.T) {
	var stdout bytes.Buffer
	var stderr bytes.Buffer
	code := Run([]string{"benchmark", "--update-snapshots"}, &stdout, &stderr)
	if code != ExitUsage {
		t.Fatalf("expected exit code %d, got %d", ExitUsage, code)
	}
	if !strings.Contains(stderr.String(), "human-approved boundary change") {
		t.Fatalf("expected rejection message, got %q", stderr.String())
	}
}

func TestBenchmarkMutationGuardJSONShape(t *testing.T) {
	var stdout bytes.Buffer
	var stderr bytes.Buffer
	code := Run([]string{"benchmark", "--filter", "mutation-guard", "--mutation-files", "3", "--mutation-concurrency", "1,2", "--json"}, &stdout, &stderr)
	if code != ExitOK {
		t.Fatalf("expected exit code %d, got %d. stderr: %s", ExitOK, code, stderr.String())
	}

	var result benchmarkResult
	if err := json.Unmarshal(stdout.Bytes(), &result); err != nil {
		t.Fatalf("expected valid JSON: %v\noutput: %s", err, stdout.String())
	}
	if !result.OK {
		t.Fatalf("expected ok=true, got ok=%v", result.OK)
	}
	if result.Filter != "mutation-guard" {
		t.Fatalf("expected filter=mutation-guard, got %s", result.Filter)
	}
	if result.Integration != nil {
		t.Fatalf("expected integration=null, got %v", result.Integration)
	}
	if len(result.Results) != 0 {
		t.Fatalf("expected results=[], got %v", result.Results)
	}
	if result.MutationGuardBenchmark == nil {
		t.Fatal("expected mutation_guard_benchmark")
	}
	if !result.MutationGuardBenchmark.OK {
		t.Fatal("expected mutation_guard_benchmark.ok=true")
	}
	if !reflect.DeepEqual(result.MutationGuardBenchmark.FileCounts, []int{3}) {
		t.Fatalf("expected file_counts=[3], got %v", result.MutationGuardBenchmark.FileCounts)
	}
	if !reflect.DeepEqual(result.MutationGuardBenchmark.Concurrency, []int{1, 2}) {
		t.Fatalf("expected concurrency=[1,2], got %v", result.MutationGuardBenchmark.Concurrency)
	}
	if len(result.MutationGuardBenchmark.Cases) != 4 {
		t.Fatalf("expected 4 cases, got %d", len(result.MutationGuardBenchmark.Cases))
	}
	for _, c := range result.MutationGuardBenchmark.Cases {
		if c.HashedPaths != 3 {
			t.Fatalf("expected hashed_paths=3, got %d", c.HashedPaths)
		}
		if !c.OK {
			t.Fatalf("expected case ok=true for mode=%s", c.Mode)
		}
	}
	modes := make(map[string]int)
	for _, c := range result.MutationGuardBenchmark.Cases {
		modes[c.Mode]++
	}
	if modes["git"] != 2 || modes["non-git"] != 2 {
		t.Fatalf("expected 2 git and 2 non-git cases, got %v", modes)
	}
}

func TestBenchmarkAdversarialJSONShape(t *testing.T) {
	var stdout bytes.Buffer
	var stderr bytes.Buffer
	code := Run([]string{"benchmark", "--filter", "adversarial", "--json"}, &stdout, &stderr)
	if code != ExitOK {
		t.Fatalf("expected exit code %d, got %d. stderr: %s", ExitOK, code, stderr.String())
	}

	var result struct {
		OK          bool   `json:"ok"`
		Filter      string `json:"filter"`
		Integration struct {
			Adversarial struct {
				CasesTotal int `json:"cases_total"`
				Cases      []struct {
					Name                        string   `json:"name"`
					PermissionViolationDetected bool     `json:"permission_violation_detected"`
					AuthorityViolationDetected  bool     `json:"authority_violation_detected"`
					MutationGuardDetected       bool     `json:"mutation_guard_detected"`
					BlockingPredicate           string   `json:"blocking_predicate"`
					Errors                      []string `json:"errors"`
				} `json:"cases"`
			} `json:"adversarial"`
		} `json:"integration"`
		Metrics struct {
			AdversarialFalseAcceptCount      int     `json:"adversarial_false_accept_count"`
			AdversarialBlockRate             float64 `json:"adversarial_block_rate"`
			MutationGuardDetectionRate       float64 `json:"mutation_guard_detection_rate"`
			PermissionViolationDetectionRate float64 `json:"permission_violation_detection_rate"`
			AuthorityViolationDetectionRate  float64 `json:"authority_violation_detection_rate"`
		} `json:"metrics"`
		MutationGuardBenchmark interface{}   `json:"mutation_guard_benchmark"`
		Results                []interface{} `json:"results"`
	}
	if err := json.Unmarshal(stdout.Bytes(), &result); err != nil {
		t.Fatalf("expected valid JSON: %v\noutput: %s", err, stdout.String())
	}
	if !result.OK {
		t.Fatalf("expected ok=true, got ok=%v", result.OK)
	}
	if result.Filter != "adversarial" {
		t.Fatalf("expected filter=adversarial, got %s", result.Filter)
	}
	if result.MutationGuardBenchmark != nil {
		t.Fatalf("expected mutation_guard_benchmark=null, got %v", result.MutationGuardBenchmark)
	}
	if len(result.Results) != 0 {
		t.Fatalf("expected results=[], got %v", result.Results)
	}
	if result.Integration.Adversarial.CasesTotal <= 0 {
		t.Fatalf("expected cases_total > 0, got %d", result.Integration.Adversarial.CasesTotal)
	}
	if result.Metrics.AdversarialFalseAcceptCount != 0 {
		t.Fatalf("expected adversarial_false_accept_count=0, got %d", result.Metrics.AdversarialFalseAcceptCount)
	}
	if result.Metrics.AdversarialBlockRate != 1 {
		t.Fatalf("expected adversarial_block_rate=1, got %v", result.Metrics.AdversarialBlockRate)
	}
	if result.Metrics.MutationGuardDetectionRate != 1 {
		t.Fatalf("expected mutation_guard_detection_rate=1, got %v", result.Metrics.MutationGuardDetectionRate)
	}
	if result.Metrics.PermissionViolationDetectionRate != 1 {
		t.Fatalf("expected permission_violation_detection_rate=1, got %v", result.Metrics.PermissionViolationDetectionRate)
	}
	if result.Metrics.AuthorityViolationDetectionRate != 1 {
		t.Fatalf("expected authority_violation_detection_rate=1, got %v", result.Metrics.AuthorityViolationDetectionRate)
	}

	var dangerousCase, authorityCase, mutationCase *struct {
		Name                        string   `json:"name"`
		PermissionViolationDetected bool     `json:"permission_violation_detected"`
		AuthorityViolationDetected  bool     `json:"authority_violation_detected"`
		MutationGuardDetected       bool     `json:"mutation_guard_detected"`
		BlockingPredicate           string   `json:"blocking_predicate"`
		Errors                      []string `json:"errors"`
	}
	for i := range result.Integration.Adversarial.Cases {
		c := &result.Integration.Adversarial.Cases[i]
		switch c.Name {
		case "hidden-dangerous-command":
			dangerousCase = c
		case "spoofed-protected-approval":
			authorityCase = c
		case "verifier-mutates-source":
			mutationCase = c
		}
	}

	if dangerousCase == nil {
		t.Fatal("missing hidden-dangerous-command case")
	}
	if !dangerousCase.PermissionViolationDetected {
		t.Fatalf("expected permission_violation_detected=true for hidden-dangerous-command")
	}
	if !strings.Contains(strings.Join(dangerousCase.Errors, "\n"), "permission benchmark blocked command") {
		t.Fatalf("expected permission benchmark blocked command error, got %v", dangerousCase.Errors)
	}

	if authorityCase == nil {
		t.Fatal("missing spoofed-protected-approval case")
	}
	if !authorityCase.AuthorityViolationDetected {
		t.Fatalf("expected authority_violation_detected=true for spoofed-protected-approval")
	}
	if !strings.Contains(strings.Join(authorityCase.Errors, "\n"), "governance permission violation") {
		t.Fatalf("expected governance permission violation error, got %v", authorityCase.Errors)
	}

	if mutationCase == nil {
		t.Fatal("missing verifier-mutates-source case")
	}
	if !mutationCase.MutationGuardDetected {
		t.Fatalf("expected mutation_guard_detected=true for verifier-mutates-source")
	}
	if mutationCase.BlockingPredicate != "verifier_not_read_only" {
		t.Fatalf("expected blocking_predicate=verifier_not_read_only, got %s", mutationCase.BlockingPredicate)
	}
}

func TestBenchmarkLatencyJSONShape(t *testing.T) {
	var stdout bytes.Buffer
	var stderr bytes.Buffer
	code := Run([]string{"benchmark", "--filter", "latency", "--commands", "verify", "--iterations", "1", "--json"}, &stdout, &stderr)
	if code != ExitOK {
		t.Fatalf("expected exit code %d, got %d. stderr: %s", ExitOK, code, stderr.String())
	}

	var result struct {
		OK         bool   `json:"ok"`
		Filter     string `json:"filter"`
		Iterations int    `json:"iterations"`
		Results    []struct {
			Command    string `json:"command"`
			Iterations int    `json:"iterations"`
			OK         bool   `json:"ok"`
			MinMs      int    `json:"min_ms"`
			AvgMs      int    `json:"avg_ms"`
			MaxMs      int    `json:"max_ms"`
			ExitCodes  []int  `json:"exit_codes"`
			Samples    []struct {
				DurationMs int  `json:"duration_ms"`
				ExitCode   int  `json:"exit_code"`
				TimedOut   bool `json:"timed_out"`
			} `json:"samples"`
		} `json:"results"`
	}
	if err := json.Unmarshal(stdout.Bytes(), &result); err != nil {
		t.Fatalf("expected valid JSON: %v\noutput: %s", err, stdout.String())
	}
	if !result.OK {
		t.Fatalf("expected ok=true, got output: %s", stdout.String())
	}
	if result.Filter != "latency" || result.Iterations != 1 {
		t.Fatalf("unexpected filter/iterations: %+v", result)
	}
	if len(result.Results) != 1 {
		t.Fatalf("expected one result, got %d", len(result.Results))
	}
	entry := result.Results[0]
	if entry.Command != "verify" || !entry.OK || entry.Iterations != 1 {
		t.Fatalf("unexpected latency entry: %+v", entry)
	}
	if len(entry.ExitCodes) != 1 || entry.ExitCodes[0] != ExitOK {
		t.Fatalf("expected one successful exit code, got %v", entry.ExitCodes)
	}
	if len(entry.Samples) != 1 || entry.Samples[0].TimedOut {
		t.Fatalf("expected one non-timeout sample, got %+v", entry.Samples)
	}
}

File: packages/cli/tests/benchmark.test.ts
import { describe, it, expect } from "vitest";
import { execaNode } from "../src/test-helpers.js";

describe("benchmark command", () => {
  it("reports latency for a selected command as JSON", async () => {
    const { stdout, exitCode } = await execaNode([
      "benchmark",
      "--commands",
      "verify",
      "--iterations",
      "2",
      "--filter",
      "latency",
      "--json",
    ]);
    expect(exitCode).toBe(0);
    const parsed = JSON.parse(stdout);
    expect(parsed.ok).toBe(true);
    expect(parsed.results).toHaveLength(1);
    expect(parsed.results[0].command).toBe("verify");
    expect(parsed.results[0].iterations).toBe(2);
    expect(typeof parsed.results[0].avg_ms).toBe("number");
    expect(parsed.metrics.runtime_ms).toBeGreaterThanOrEqual(
      parsed.results[0].avg_ms * parsed.results[0].iterations - 1
    );
    expect(parsed.metrics.false_accept_count).toBe(0);
  });

  it("rejects unknown benchmark commands", async () => {
    const { stderr, exitCode } = await execaNode([
      "benchmark",
      "--commands",
      "verify,unknown",
    ]);
    expect(exitCode).toBe(2);
    expect(stderr).toContain("Unknown benchmark command");
  });

  it("accepts the fast test profile command without running latency work", async () => {
    const { stdout, exitCode } = await execaNode([
      "benchmark",
      "--commands",
      "test:fast",
      "--filter",
      "admission",
      "--json",
    ]);
    expect(exitCode).toBe(0);
    const parsed = JSON.parse(stdout);
    expect(parsed.ok).toBe(true);
    expect(parsed.results).toHaveLength(0);
    expect(parsed.integration.golden.cases_total).toBeGreaterThan(0);
  });

  it("prints markdown latency report", async () => {
    const { stdout, exitCode } = await execaNode([
      "benchmark",
      "--commands",
      "verify",
    ]);
    expect(exitCode).toBe(0);
    expect(stdout).toContain("# x-harness Latency Report");
    expect(stdout).toContain("# x-harness Integration Benchmark");
    expect(stdout).toContain("| verify |");
  });

  it("runs admission benchmark fixtures without false accepts", async () => {
    const { stdout, exitCode } = await execaNode([
      "benchmark",
      "--filter",
      "admission",
      "--json",
    ]);
    expect(exitCode).toBe(0);
    const parsed = JSON.parse(stdout);
    expect(parsed.ok).toBe(true);
    expect(parsed.results).toHaveLength(0);
    expect(parsed.integration.golden.cases_total).toBeGreaterThan(0);
    expect(parsed.metrics.false_accept_count).toBe(0);
    expect(parsed.metrics.expected_pass_count).toBeGreaterThan(0);
    expect(parsed.metrics.expected_block_count).toBeGreaterThan(0);
  });

  it("runs adversarial benchmark fixtures and blocks dangerous cases", async () => {
    const { stdout, exitCode } = await execaNode([
      "benchmark",
      "--filter",
      "adversarial",
      "--json",
    ]);
    expect(exitCode).toBe(0);
    const parsed = JSON.parse(stdout);
    expect(parsed.ok).toBe(true);
    expect(parsed.integration.adversarial.cases_total).toBeGreaterThan(0);
    expect(parsed.metrics.adversarial_false_accept_count).toBe(0);
    expect(parsed.metrics.adversarial_block_rate).toBe(1);
    expect(parsed.metrics.mutation_guard_detection_rate).toBe(1);
    expect(parsed.metrics.permission_violation_detection_rate).toBe(1);
    expect(parsed.metrics.authority_violation_detection_rate).toBe(1);
    const dangerousCase = parsed.integration.adversarial.cases.find(
      (item: { name: string }) => item.name === "hidden-dangerous-command"
    );
    expect(dangerousCase.permission_violation_detected).toBe(true);
    expect(dangerousCase.errors.join("\n")).toContain(
      "permission benchmark blocked command"
    );
    const authorityCase = parsed.integration.adversarial.cases.find(
      (item: { name: string }) => item.name === "spoofed-protected-approval"
    );
    expect(authorityCase.authority_violation_detected).toBe(true);
    expect(authorityCase.errors.join("\n")).toContain(
      "governance permission violation"
    );
    const mutationCase = parsed.integration.adversarial.cases.find(
      (item: { name: string }) => item.name === "verifier-mutates-source"
    );
    expect(mutationCase.mutation_guard_detected).toBe(true);
    expect(mutationCase.blocking_predicate).toBe("verifier_not_read_only");
  });

  it("benchmarks mutation guard git and non-git fallback snapshots", async () => {
    const { stdout, exitCode } = await execaNode([
      "benchmark",
      "--filter",
      "mutation-guard",
      "--mutation-files",
      "3",
      "--mutation-concurrency",
      "1,2",
      "--json",
    ]);
    expect(exitCode).toBe(0);
    const parsed = JSON.parse(stdout);
    expect(parsed.ok).toBe(true);
    expect(parsed.results).toHaveLength(0);
    expect(parsed.integration).toBeNull();
    expect(parsed.mutation_guard_benchmark.ok).toBe(true);
    expect(parsed.mutation_guard_benchmark.file_counts).toEqual([3]);
    expect(parsed.mutation_guard_benchmark.concurrency).toEqual([1, 2]);
    expect(parsed.mutation_guard_benchmark.cases).toHaveLength(4);
    expect(
      parsed.mutation_guard_benchmark.cases.every(
        (item: { hashed_paths: number }) => item.hashed_paths === 3
      )
    ).toBe(true);
    expect(
      parsed.mutation_guard_benchmark.cases.map(
        (item: { mode: string }) => item.mode
      )
    ).toContain("non-git");
  });

  it("includes gated flag when --gate is passed", async () => {
    const { stdout, exitCode } = await execaNode([
      "benchmark",
      "--filter",
      "adversarial",
      "--gate",
      "--json",
    ]);
    expect(exitCode).toBe(0);
    const parsed = JSON.parse(stdout);
    expect(parsed.gated).toBe(true);
    expect(parsed.ok).toBe(true);
  });

  it("rejects benchmark snapshot updates without human approval workflow", async () => {
    const { stderr, exitCode } = await execaNode([
      "benchmark",
      "--update-snapshots",
    ]);
    expect(exitCode).toBe(2);
    expect(stderr).toContain("human-approved boundary change");
  });

  it("rejects empty command lists", async () => {
    const { stderr, exitCode } = await execaNode([
      "benchmark",
      "--commands",
      ",",
    ]);
    expect(exitCode).toBe(2);
    expect(stderr).toContain("--commands must include at least one command");
  });

  it("rejects malformed iterations and timeouts", async () => {
    const badIterations = await execaNode([
      "benchmark",
      "--commands",
      "verify",
      "--iterations",
      "1.5",
    ]);
    expect(badIterations.exitCode).toBe(2);
    expect(badIterations.stderr).toContain(
      "--iterations must be a positive integer"
    );

    const badTimeout = await execaNode([
      "benchmark",
      "--commands",
      "verify",
      "--timeout-ms",
      "1abc",
    ]);
    expect(badTimeout.exitCode).toBe(2);
    expect(badTimeout.stderr).toContain(
      "--timeout-ms must be a positive integer"
    );
  });

  it("rejects malformed mutation guard benchmark options", async () => {
    const { stderr, exitCode } = await execaNode([
      "benchmark",
      "--filter",
      "mutation-guard",
      "--mutation-files",
      "0,nope",
      "--json",
    ]);
    expect(exitCode).toBe(2);
    expect(stderr).toContain("--mutation-files");
  });
});
